### PR TITLE
WIP: lightningd/plugin: add field `early_conf`, allow plugin early (sync) config or abort

### PR DESCRIPTION
**Pull request by SimonVrouwe**
Another proposal to address issues raised in #2988, but now add a new plugins category `early_conf`. Plugins that set `early_conf = True` in `getmanifest` will initialize early in lightningd's startup, i.e. before the database is touched.

Useful to make a backup [plugin](https://github.com/lightningd/plugins/pull/40) more bulletproof. The  plugin now doesn't need to worry about db writes that happen before `init` or have to think about what their importance may be when crashing. It can now also safely copy the db during `init` without missing any `PRAGMA` statements (see #3007) or other that doesn't persist in the db (file).

Additionally this PR allows plugin `init` to return an `error`, which is handled as follow:
- At startup (meaning: not started via rpc `plugin start`) an init-error from `early_conf` plugin kills lightningd and aborts lightningd startup early.
- Outside startup (when started via `plugin start` cmd) an init-error from the plugin (any type) makes lightningd kill the plugin.

So `early_conf` plugins can now shutdown `lightningd` before anything happened, this wasn't possible before without hacks.

~Plugins that don't like to be started via rpc `plugin start` can check the configuration['startup'] field and return an init-error and be killed. However, plugins that register the (sync) `db_write` hook (which happens _before_ `init` is called) should then just _pass_ the hook calls when not fully initialized, or maybe it's `manifest` can check `startup = False` field and then not register the hook.~
**edit:** Plugins that are _freshly_ started via rpc `plugin start` and have set `dynamic=False` or `early_conf=True` in their manifest, should probably be killed by lightningd ASAP (in `plugin_manifest_cb`), so that no hook etc. is registered.

TODO:
- [ ] complete json error handling
- [ ] add tests
- [ ] update docs

### CODE REVIEW

**Comment by darosior** [comment]
> Plugins that are freshly started via rpc plugin start and have set dynamic=False or early_conf=True in their manifest, should probably be killed by lightningd ASAP

Why ? This is a common case where a plugin doesnt want to be stopped before `lightningd` is stopped but can be started at any time.

**Comment by SimonVrouwe** [comment]
> This is a common case

@darosier Do  you have an example of that?  Perhaps I am confused about the (intended) meaning of  `dynamic` here, as trying to stop a static plugin returns "plugin cannot be _managed_ when lightningd is up", well _managed_ sounds as that it includes _starting_, I guess?

**Comment by darosior** [comment]
Actually I originally wanted to show this error for both `start` and `stop` subcommands : but it finally seemed useful to start plugins after `lightningd` startup. Maybe we should change the error. Cf https://github.com/ElementsProject/lightning/blob/49496ab3bac15e2ad00f181e92b8986ada5d9a89/lightningd/plugin_control.c#L29

**Comment by SimonVrouwe** [comment]
> originally wanted to show this error for both start and stop subcommands

But then you first need to start it and wait for  `getmanifest` to return its `dynamic` variable, which would change the design of `plugin` command, i.e. it should then probably return a `command_still_pending`.

@darosior BTW you still haven't told me your case of a plugin that for some reason does not want to be stopped via RPC (`dynamic=False`) but has no problem being started via RPC.

I was also thinking, why are these decisions (start/stop policies etc.) not delegated to the plugin itself by adding a `stop` call in addition to `getmanifest` and `init`? And then have the plugin decide what to do, which (again) would require your `plugin` command return a `command_still_pending`.

Then finally, if the only motivation for this PR is for a plugin that uses the `db_write` hook to create  a backup, then maybe it shouldn't be a _plugin_ because (apparently) it just doesn't fit into that category well enough.

**Comment by darosior** [comment]
> > originally wanted to show this error for both start and stop subcommands
> 
> But then you first need to start it and wait for `getmanifest` to return its `dynamic` variable, which would change the design of `plugin` command, i.e. it should then probably return a `command_still_pending`.

This could indeed have been a pending command until the plugin is either started or fails, but simplicity was prefered.

> @darosior BTW you still haven't told me your case of a plugin that for some reason does not want to be stopped via RPC (`dynamic=False`) but has no problem being started via RPC.

Actually, I answered you just above....

> I was also thinking, why are these decisions (start/stop policies etc.) not delegated to the plugin itself by adding a `stop` call in addition to `getmanifest` and `init`? And then have the plugin decide what to do

I think that's something we want.

> , which (again) would require your `plugin` command return a `command_still_pending`.

As mentioned above, simplicity was prefered when integrated, but feel free to improve and make the `plugin` command (which is not mine..) a `command_still_pending` : I'd be happy to review.

**Comment by darosior** [comment]
> Then finally, if the only motivation for this PR is for a plugin that uses the db_write hook to create a backup, then maybe it shouldn't be a plugin because (apparently) it just doesn't fit into that category well enough.

> Or maybe a plugin is just not the right choice for this or even having a sync db_write hook for plugins was a mistake because it just doesn't fit in well enough.
*(Quote from https://github.com/lightningd/plugins/pull/40)*

Maybe we should just introduce a notification about the DB writes since a backup plugin doesnt need to take control over this writes ? I think we should wait for #2924 if we do this.

**Comment by cdecker** [comment]
Really like the idea of giving the plugin an early warning that it is about to receive notifications and hook calls, but maybe we should make it a `preinit` instread of moving the `init` call?

My main fear is that we are introducing more states into the process, i.e., some plugins get inited early, some late. imho a plugin should be able to signal in the `getmanifest` that it requires a pre-init with some info (lightning-dir) before any other notification or hook call.

I would just make it a `preinit` notification that is delivered just after we `cd` into the lightning-dir and before even opening the DB or starting the `io_loop`. We can make it a notification since it is optional (we decided to make `init` an RPC call in order not to force plugin writers to have to implement notification handling if they didn't need to. Since this is about notifications and hook calls anyway, this rationale is not applicable here). Alternatively we can even make it a hook call, that'd allow us to even kill the main daemon if something goes terribly wrong during startup.

How does that sound? imho it minimizes code movements, plugin state changes, and is nicely backwards compatible, while still accomplishing the goal.

> Maybe we should just introduce a notification about the DB writes since a backup plugin doesnt need to take control over this writes ? I think we should wait for #2924 if we do this.

The key with the `db_write` hook is that it is called synchronously and therefore we can ensure the write was backed up before continuing to make progress. If we were to just use notifications, the plugin may not have stored the write in the backup when it gets killed, resulting in the backup and the DB being out of sync.

**Comment by SimonVrouwe** [comment]
> I would just make it a preinit notification that is delivered just after we cd into the lightning-dir and before even opening the DB

If the notification includes the `configuration` field (same as in `init`) then this looks useful. So it should be send after [handle_opts](https://github.com/ElementsProject/lightning/blob/0e2ec275769b2bb170bb81a39bb6f8fd3171d015/lightningd/lightningd.c#L681) and before [wallet_new](https://github.com/ElementsProject/lightning/blob/0e2ec275769b2bb170bb81a39bb6f8fd3171d015/lightningd/lightningd.c#L693).

If I understand correctly, the preinit notification is a sort of fake `init` signal for the plugin to prepare (check options, db files etc.) and be ready for the first `db_write` hook call. Assumed the preinit notification is handled before the first `db_write` hook call, the plugin can block `lightningd`'s [startup](https://github.com/ElementsProject/lightning/blob/0e2ec275769b2bb170bb81a39bb6f8fd3171d015/lightningd/plugin.c#L1130), until the plugin has completed pre-init. The pre-init could set some flag to tell the `db_write` hook handler to return False and kill the daemon if it wants to abort the startup.

Sounds good. And what about plugins that shouldn't be started at run-time via RPC? Can they return an error in `getmanifest` (when `startup=False`) and be killed before registering their sync hooks?

**Comment by cdecker** [comment]
> Assumed the preinit notification is handled before the first `db_write` hook call, the plugin can block `lightningd`'s [startup](https://github.com/ElementsProject/lightning/blob/0e2ec275769b2bb170bb81a39bb6f8fd3171d015/lightningd/plugin.c#L1130), until the plugin has completed pre-init. The pre-init could set some flag to tell the `db_write` hook handler to return False and kill the daemon if it wants to abort the startup.

Just one caveat: notifications are non-blocking, but they will be delivered in-order. So it is not possible for a notification to hold up startup (and I don't think it should), but it can process the notification and is assured to receive the signals in the correct order.

> Sounds good. And what about plugins that shouldn't be started at run-time via RPC? Can they return an error in `getmanifest` (when `startup=False`) and be killed before registering their sync hooks?

I thought that was the `dynamic` flag: if `dynamic` is set they can be controlled by the `plugin` command, otherwise not. If a plugin requires guarantees on notification delivery and getting all hook calls it should not be marked as `dynamic`, but we can leave that to plugin authors.

**Comment by SimonVrouwe** [comment]
> Assumed the preinit notification is handled before the first db_write hook call, the plugin can block lightningd's startup, until the plugin has completed pre-init.

Meaning it can _effectively_ hold up the `lightningd` startup because, while the (single thread) plugin handles the notification, it cannot handle the hook for which `lightningd` is waiting. If the plugin wants to abort the startup, it can make the hook handler return false. Which is also a bit hack-ish.

**TLDR**; It should probably be a **preinit hook instead of notification**, that gives plugins direct means to hold-up the startup.

> if dynamic is set they can be controlled by the plugin command, otherwise not

Yes, but a plugin needs to be started first to _know_ its preference from its manifest :-)

**Comment by SimonVrouwe** [comment]
I am having second thoughts while trying to code-up a `preinit_hook`. What should be done with the `plugin->plugin_state` around this? Using a hook or notification to (fake) `init` a plugin actually seems to mess up the plugin's state.

> My main fear is that we are introducing more states into the process, i.e., some plugins get inited early, some late.

What exactly do you mean and what is wrong with plugins completing their init in an early stage instead of _somewhere_ in the main loop?
